src/solvers/refinement/string_refinement.cpp

```diff
@@ -257,18 +257,22 @@ static std::vector<exprt> generate_instantiations(
 /// \param expr: an expression possibly containing function applications
 /// \param generator: generator for the string constraints
 /// \return an expression containing no function application
-static exprt substitute_function_applications(
-  exprt expr,
+static void substitute_function_applications(
+  exprt &expr,
   string_constraint_generatort &generator)
 {
-  for(auto &operand : expr.operands())
-    operand = substitute_function_applications(operand, generator);
-
-  if(expr.id() == ID_function_application)
-    return generator.add_axioms_for_function_application(
-      to_function_application_expr(expr));
-
-  return expr;
+  for(auto it = expr.depth_begin(); it != expr.depth_end();)
```

Comment:
1. See the other commit: I'd appreciate a `// no ++it` comment
2. Why does the other commit make an effort of storing the end iterator up front, while that's not done here. As `depth_end()` can't be inlined, it may be advisable to do it here as well.